server/src/test/java/org/elasticsearch/search/functionscore/QueryRescorerIT.java

```diff
@@ -748,4 +754,55 @@ public void testRescorePhaseWithInvalidSort() throws Exception {
             assertThat(hit.getScore(), equalTo(101f));
         }
     }
+
+    public void testRescoreAfterCollapse() throws Exception {
+        assertAcked(prepareCreate("test")
+            .addMapping(
+                "type1",
+                jsonBuilder()
+                .startObject()
+                .startObject("properties")
+                .startObject("group")
+                .field("type", "keyword")
+                .endObject()
+                .endObject()
+                .endObject())
+        );
+
+        indexDocument(1, "value", "a");
+        indexDocument(2, "one one value", "a");
+        indexDocument(3, "one one two value", "b");
+        // should be highest on rescore, but filtered out during collapse
+        indexDocument(4, "one two two value", "b");
+
+        refresh("test");
+
+        SearchResponse searchResponse = client().prepareSearch("test")
+            .setTypes("type1")
+            .setQuery(new MatchQueryBuilder("name", "one"))
```

Comment:
The score of this query depends on the number of shards, the default similarity, ... To make sure that we have consistent scoring you can use a `function_score` query like the following:
````
 QueryBuilder query = functionScoreQuery(
            termQuery("name", "one"),
            ScoreFunctionBuilders.fieldValueFactorFunction("my_static_doc_score") 
        ).boostMode(CombineFunction.REPLACE);
````
... and add the `my_static_doc_score` at indexing time.

Reply:
fixed